sql/core/src/main/scala/org/apache/spark/sql/ForeachWriter.scala

```diff
@@ -50,14 +50,13 @@ import org.apache.spark.annotation.Evolving
  *
  * Important points to note:
  * <ul>
- * <li>The `partitionId` and `epochId` can be used to deduplicate generated data when failures
- *     cause reprocessing of some input data. This depends on the execution mode of the query. If
- *     the streaming query is being executed in the micro-batch mode, then every partition
- *     represented by a unique tuple (partitionId, epochId) is guaranteed to have the same data.
- *     Hence, (partitionId, epochId) can be used to deduplicate and/or transactionally commit data
- *     and achieve exactly-once guarantees. However, if the streaming query is being executed in the
- *     continuous mode, then this guarantee does not hold and therefore should not be used for
- *     deduplication.
+ * <li>Spark doesn't guarantee same output for (partitionId, epochId) on failure, so deduplication
+ *     cannot be achieved with (partitionId, epochId). Refer SPARK-28650 for more details.
+ *
+ *     You can still apply deduplication on `epochId`, but there's less benefit to leverage this,
```

Comment:
Nit: I'd keep the voice consistent. "`epochId` can still be used for deduplication .." instead of "you can ..."
Can we clarify if there are cases where deduplication on both is still valid?

Reply:
> Nit: I'd keep the voice consistent. "epochId can still be used for deduplication .." instead of "you can ..."

I'll do the update.

> Can we clarify if there are cases where deduplication on both is still valid?

As you could see the cases in SPARK-28650, Spark and source can break this guarantee which end users would be not easy to determine. Guarantee can still be broken even end users don't change the query, so I'd rather not enumerating the cases and let end users encounter the odd cases. What do you think about this @zsxwing ?